allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py

```diff
@@ -29,76 +31,155 @@ class PytorchSeq2SeqWrapper(Seq2SeqEncoder):
-        packed_sequence_output, _ = self._module(packed_sequence_input, hidden_state)
-        unpacked_sequence_tensor, _ = pad_packed_sequence(packed_sequence_output, batch_first=True)
+        # Prepare the initial states.
+        if not self._stateful:
+            initial_states = hidden_state
+        else:
+            initial_states = self._get_initial_states(num_valid)
 
-        # We sorted by length, so if there are invalid rows that need to be zeroed out
-        # they will be at the end.
+        # Actually call the module on the sorted PackedSequence.
+        packed_sequence_output, final_states = self._module(packed_sequence_input, initial_states)
+        if self._stacked:
+            # packed_sequence_output is shape (n_layers, batch_size, n_times, nx)
+            num_layers = packed_sequence_output.size()[0]
+            unpacked_sequence_tensor = [
+                    layer.squeeze(0) for layer in packed_sequence_output.chunk(num_layers, 0)
+            ]
+        else:
+            # just one layer
+            num_layers = 1
+            unpacked_sequence_tensor = [pad_packed_sequence(packed_sequence_output, batch_first=True)[0]]
+
+        # Add back invalid rows.
         if num_valid < batch_size:
-            unpacked_sequence_tensor[num_valid:, :, :] = 0.
+            _, length, dim = unpacked_sequence_tensor[0].size()
+            zeros = unpacked_sequence_tensor[0].data.new(batch_size - num_valid, length, dim).fill_(0)
+            for k in range(num_layers):
+                unpacked_sequence_tensor[k] = torch.cat([unpacked_sequence_tensor[k], zeros], 0)
 
         # It's possible to need to pass sequences which are padded to longer than the
         # max length of the sequence to a Seq2SeqEncoder. However, packing and unpacking
         # the sequences mean that the returned tensor won't include these dimensions, because
         # the RNN did not need to process them. We add them back on in the form of zeros here.
-        sequence_length_difference = total_sequence_length - unpacked_sequence_tensor.size(1)
+        sequence_length_difference = total_sequence_length - unpacked_sequence_tensor[0].size(1)
         if sequence_length_difference > 0:
-            zeros = unpacked_sequence_tensor.data.new(batch_size, sequence_length_difference,
-                                                      unpacked_sequence_tensor.size(-1)).fill_(0)
+            zeros = unpacked_sequence_tensor[0].data.new(batch_size, sequence_length_difference,
+                                                         unpacked_sequence_tensor[0].size(-1)).fill_(0)
             zeros = torch.autograd.Variable(zeros)
-            unpacked_sequence_tensor = torch.cat([unpacked_sequence_tensor, zeros], 1)
+            for k in range(num_layers):
+                unpacked_sequence_tensor[k] = torch.cat([unpacked_sequence_tensor[k], zeros], 1)
 
+        if self._stateful:
+            self._update_states(final_states, num_valid)
```

Comment:
These need to be restored back to their original ordering before being set as the states, because they are the states for the sorted sequences.

Reply:
I'll do that in an upcoming commit.